lib/rex/io/socket_abstraction.rb

```diff
@@ -4,202 +4,216 @@
-  #
-  def initialize_abstraction
-    self.lsock, self.rsock = Rex::Compat.pipe
-  end
+      #
+      # Override this method to init the abstraction
+      #
+      def initialize_abstraction
+        self.lsock, self.rsock = Rex::Compat.pipe
+      end
 
-  #
-  # This method cleans up the abstraction layer.
-  #
-  def cleanup_abstraction
-    self.lsock.close if (self.lsock and !self.lsock.closed?)
-    self.rsock.close if (self.rsock and !self.rsock.closed?)
+      #
+      # This method cleans up the abstraction layer.
+      #
+      def cleanup_abstraction
+        lsock.close if lsock and !lsock.closed?
 
-    self.lsock = nil
-    self.rsock = nil
-  end
+        monitor_thread.join if monitor_thread&.alive?
 
-  #
-  # Low-level write to the local side.
-  #
-  def syswrite(buffer)
-    lsock.syswrite(buffer)
-  end
+        rsock.close if rsock and !rsock.closed?
 
-  #
-  # Low-level read from the local side.
-  #
-  def sysread(length)
-    lsock.sysread(length)
-  end
+        self.lsock = nil
+        self.rsock = nil
+      end
 
-  #
-  # Shuts down the local side of the stream abstraction.
-  #
-  def shutdown(how)
-    lsock.shutdown(how)
-  end
+      #
+      # Low-level write to the local side.
+      #
+      def syswrite(buffer)
+        lsock.syswrite(buffer)
+      end
 
-  #
-  # Closes both sides of the stream abstraction.
-  #
-  def close
-    cleanup_abstraction
-    super
-  end
+      #
+      # Low-level read from the local side.
+      #
+      def sysread(length)
+        lsock.sysread(length)
+      end
 
-  #
-  # Symbolic peer information.
-  #
-  def peerinfo
-    "Remote-side of Pipe"
-  end
+      #
+      # Shuts down the local side of the stream abstraction.
+      #
+      def shutdown(how)
+        lsock.shutdown(how)
+      end
 
-  #
-  # Symbolic local information.
-  #
-  def localinfo
-    "Local-side of Pipe"
-  end
+      #
+      # Closes both sides of the stream abstraction.
+      #
+      def close
+        cleanup_abstraction
+        super
+      end
+
+      #
+      # Symbolic peer information.
+      #
+      def peerinfo
+        'Remote-side of Pipe'
+      end
+
+      #
+      # Symbolic local information.
+      #
+      def localinfo
+        'Local-side of Pipe'
+      end
 
-  #
-  # The left side of the stream.
-  #
-  attr_reader :lsock
-  #
-  # The right side of the stream.
-  #
-  attr_reader :rsock
-
-protected
-
-  def monitor_rsock(threadname = "SocketMonitorRemote")
-    self.monitor_thread = Rex::ThreadFactory.spawn(threadname, false) {
-      loop do
-        closed = false
-        buf    = nil
-
-        if not self.rsock
-          wlog("monitor_rsock: the remote socket is nil, exiting loop")
-          break
+      #
+      # The left side of the stream.
+      #
+      attr_reader :lsock
+      #
+      # The right side of the stream.
+      #
+      attr_reader :rsock
+
+      module MonitoredRSock
+        def close
+          @close_requested = true
```

Comment:
This only ever goes from true to false so I don't think there's a synchronization problem here but if another solution is preferred happy to go a different direction.